### PR TITLE
EA Popular comments algorithm updates

### PR DESCRIPTION
**Pull request by oetherington**
This PR implementes the discussed changes for the popular comments algorithm:

- Only 1 popular comment per post
- Multiply downvote weight by 3

To calculate the score I added some SQL functions to calculate Wilson score intervals with an optional downvote multiplier.

There's also some early-stage infrastucture for adding paginated custom resolvers based on `useMulti`. 

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1205160535144996) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by s-cheng** [comment]
@oetherington A couple small things you might want to look into, before I finish reviewing:
1. Upon clicking "load more", sometimes the original list of comments changes. (You can see in the example below, that the third comment moved down.) This isn't a matter of data changing, since if I refresh then I see the original three comments again.

https://github.com/ForumMagnum/ForumMagnum/assets/9057804/3c83dda8-d966-42df-9dcb-1f14c1f0402f

2. I think the algorithm isn't quite getting the right comments. If I compare this branch with prod, this branch gets a comment from [this post](https://forum.effectivealtruism.org/posts/Bjr6FXvnKqb37uMPP/shutting-down-ai-safety-support) that has a much lower score than prod does. The difference can't be explained by the downvote multiplier (I don't want to say more here but you can look at the downvotes in the db).

This branch:
Screen Shot 2023-08-01 at 2 56 49 PM

Prod:
Screen Shot 2023-08-01 at 2 57 05 PM

File: packages/lesswrong/server/sqlConnection.ts [hunk 1]
```diff
@@ -193,6 +193,60 @@ const onConnectQueries: string[] = [
...
+          -SUM(v."power") FILTER (WHERE v."voteType" IN ('bigDownvote', 'smallDownvote')),
+          0
+        )::INTEGER,
+        downvote_multiplier
+      )
+    FROM "Comments" c
+    JOIN "Votes" v ON
+      v."documentId" = c."_id" AND
+      v."isUnvote" IS NOT TRUE AND
+      v."cancelled" IS NOT TRUE
```
Comment (s-cheng) [inline packages/lesswrong/server/sqlConnection.ts:246]: Maybe also check for `collectionName` = `Comments`? Not sure if that would make a difference

File: packages/lesswrong/server/repos/CommentsRepo.ts [hunk 2]
```diff
@@ -75,6 +74,32 @@ export default class CommentsRepo extends AbstractRepo<DbComment> {
...
+          CURRENT_TIMESTAMP - "postedAt" < '1 week'::INTERVAL AND
+          "shortform" IS NOT TRUE AND
+          "baseScore" >= $1 AND
+          "retracted" IS NOT TRUE AND
+          "deleted" IS NOT TRUE AND
+          "deletedPublic" IS NOT TRUE AND
+          "needsReview" IS NOT TRUE
+        ORDER BY "postId"
+      ) c
+      ORDER BY fm_comment_confidence(c."_id", 3) DESC
```
Comment (s-cheng) [inline packages/lesswrong/server/repos/CommentsRepo.ts:97]: Does `fm_comment_confidence()` get called for every comment every time this is run?

**Review by s-cheng — COMMENTED** [review]
Some exciting additions here, thanks! ^^ I'd like to review this again after the two issues I mentioned earlier are addressed

File: packages/lesswrong/server/repos/CommentsRepo.ts (hunk 2 above)
Comment (oetherington) [reply packages/lesswrong/server/repos/CommentsRepo.ts]: Only the comments that are returned by the sub-query, which is generally ~40. I was planning to add a cache on top of this too in a separate PR before rolling out to all users, but it was extremely simple so I just went ahead and added it here.

File: packages/lesswrong/components/hooks/usePaginatedResolver.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  let results = data?.[resolverName]?.results;
+  if (results && results.length > limit) {
+    results = take(results, limit);
+  }
+
+  useEffect(() => {
+    if (results?.length && results.length > (list?.length ?? 0)) {
+      setList(results);
+    }
+  }, [results, list?.length]);
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/components/hooks/usePaginatedResolver.ts:100]: There was some weird behaviour with the graphql query getting invalidated when clicking load more. JP and I spent a while looking at it but it didn't seem worth investing much time in. Doing it this way also makes it easier to implement adding an offset in the future if we want.

**Comment by oetherington** [comment]
@s-cheng Bump on this

**Review by s-cheng — APPROVED** [review]
🦭

**Comment by jpaddison3** [comment]
What's the plan here? JP's guess:

* Deploy this
* Deploy a change to let mods mark posts as no-fly-zones
* Deploy a change to make it go live to beta users

cc @SharangP

**Comment by oetherington** [comment]
I was about to deploy literally now (will hold off pending answers above now though). This PR already includes enabling it for beta users. I was planning to add filtering out posts as a separate PR in the next day or two (should be very simple), but maybe you'd prefer to include it directly here?

**Comment by jpaddison3** [comment]
Ah, I see. Before you deploy could you add some posts to the exclusion db setting (I'll DM)? Otherwise, sounds good to deploy.

**Comment by oetherington** [comment]
@jpaddison3 Ok, cool. I just went ahead and added the field to posts since we're about to delete the database setting.

**Pull request merged**